File: web/timeline/references.js
import {
  ASSET_SOURCE_FILE_PATH,
  ASSET_TYPE_IMAGE,
  deepClone,
} from "./schema.js";

export const REFERENCE_KIND_CHARACTER = "character";
export const PROMPT_REFERENCE_TRIGGER = "@";

const REFERENCE_TAG_RE = /@(?<label>image[1-9]\d*):(?<kind>[A-Za-z][A-Za-z0-9_-]*)(?:\[(?<strength>[^\]]*)\])?/g;

export function getCharacterReferences(timeline) {
  const references = timeline?.project?.metadata?.character_references;
  return Array.isArray(references) ? references : [];
}

export function activeCharacterReferences(timeline) {
  return getCharacterReferences(timeline).filter((reference) => reference.enabled !== false);
}

export function getReferencePromptCompletions(timeline) {
  return activeCharacterReferences(timeline).map((reference) => ({
    id: reference.id,
    label: reference.label,
    tag: formatCharacterReferenceTag(reference),
    description: reference.description || "",
    trigger: PROMPT_REFERENCE_TRIGGER,
  }));
}

export function formatCharacterReferenceTag(referenceOrLabel, strengthOverride = null) {
  const label = normalizeReferenceLabel(
    typeof referenceOrLabel === "string" ? referenceOrLabel : referenceOrLabel?.label,
  );
  const strength = finiteNumberOrNull(strengthOverride);
  return strength == null
    ? `@${label}:${REFERENCE_KIND_CHARACTER}`
    : `@${label}:${REFERENCE_KIND_CHARACTER}[${strength}]`;
}

export function parseReferenceTags(prompt) {
  const tags = [];
  const text = String(prompt ?? "");
  for (const match of text.matchAll(REFERENCE_TAG_RE)) {
    const label = normalizeReferenceLabel(match.groups?.label);
    const kind = String(match.groups?.kind ?? "").toLowerCase();
    const strengthOverride = finiteNumberOrNull(match.groups?.strength);
    tags.push({
      label,
      kind,
      token: match[0],
      supported: kind === REFERENCE_KIND_CHARACTER,
      strength_override: strengthOverride,
    });
  }
  return tags;
}

export function normalizeReferenceLabel(value, fallbackIndex = 0) {
  const label = String(value ?? "").trim().toLowerCase();
  return /^image[1-9]\d*$/.test(label) ? label : `image${Number(fallbackIndex) + 1}`;
}

export function normalizeCharacterReferences(value) {
  if (!Array.isArray(value)) return [];
  return value
    .filter((reference) => reference && typeof reference === "object" && !Array.isArray(reference))
    .map((reference, index) => {
      const normalized = deepClone(reference);
      normalized.label = normalizeReferenceLabel(normalized.label, index);
      normalized.id = String(normalized.id || normalized.label);
      normalized.kind = REFERENCE_KIND_CHARACTER;
      normalized.enabled = normalized.enabled !== false;
      normalized.description = String(normalized.description ?? "");
      normalized.strength = clampStrength(normalized.strength);
      normalized.image = normalizeReferenceImage(normalized.image);
      return normalized;
    });
}

export function ensureCharacterReferences(timeline) {
  timeline.project ??= {};
  timeline.project.metadata = timeline.project.metadata && typeof timeline.project.metadata === "object" && !Array.isArray(timeline.project.metadata)
    ? timeline.project.metadata
    : {};
  timeline.project.metadata.character_references = normalizeCharacterReferences(timeline.project.metadata.character_references);
  return timeline.project.metadata.character_references;
}

export function createCharacterReferenceFromPickedItem(timeline, item) {
  const image = createReferenceImageFromPickedItem(item);
  if (!image) return null;
  const references = getCharacterReferences(timeline);
  const label = nextReferenceLabel(references);
  return {
    id: makeReferenceId(label, image.path || image.name || ""),
    label,
    kind: REFERENCE_KIND_CHARACTER,
    enabled: true,
    description: "",
    strength: 1.0,
    image,
  };
}

export function addCharacterReference(timeline, item) {
  const references = ensureCharacterReferences(timeline);
  const reference = createCharacterReferenceFromPickedItem(timeline, item);
  if (!reference) return null;
  references.push(reference);
  return reference;
}

export function removeCharacterReference(timeline, referenceId) {
  const references = ensureCharacterReferences(timeline);
  const next = references.filter((reference) => reference.id !== referenceId);
  timeline.project.metadata.character_references = next;
  return next.length !== references.length;
}

export function createReferenceImageFromPickedItem(item) {
  const path = String(item?.path ?? "").trim();
  if (!path) return null;
  return normalizeReferenceImage({
    type: ASSET_TYPE_IMAGE,
    source_kind: ASSET_SOURCE_FILE_PATH,
    path,
    name: item.name ?? basename(item.filename ?? path),
    mime_type: item.mime_type ?? "",
    size_bytes: Number.isFinite(item.size) ? item.size : null,
    metadata: {
      browser_alias: item.folder_alias ?? null,
      browser_filename: item.filename ?? null,
      mtime: Number.isFinite(item.mtime) ? item.mtime : null,
      width: Number.isFinite(item.width) ? item.width : null,
      height: Number.isFinite(item.height) ? item.height : null,
    },
  });
}

export function normalizeReferenceImage(image) {
  if (!image || typeof image !== "object" || Array.isArray(image)) return null;
  const path = String(image.path ?? image.file_path ?? "").trim();
  const normalized = deepClone(image);
  normalized.type = ASSET_TYPE_IMAGE;
  normalized.source_kind = normalized.source_kind || ASSET_SOURCE_FILE_PATH;
  normalized.path = path || null;
  normalized.name = String(normalized.name ?? basename(path) ?? "");
  normalized.mime_type = normalized.mime_type ?? "";
  normalized.size_bytes = Number.isFinite(normalized.size_bytes) ? normalized.size_bytes : null;
  normalized.metadata = normalized.metadata && typeof normalized.metadata === "object" && !Array.isArray(normalized.metadata)
    ? normalized.metadata
    : {};
  delete normalized.asset_id;
  return normalized;
}

function nextReferenceLabel(references) {
  const used = new Set(references.map((reference, index) => normalizeReferenceLabel(reference?.label, index)));
  let index = 1;
  while (used.has(`image${index}`)) index += 1;
  return `image${index}`;
}

function clampStrength(value) {
  const numeric = Number(value);
  if (!Number.isFinite(numeric)) return 1.0;
  return Math.max(0, Math.min(1, numeric));
}

function finiteNumberOrNull(value) {
  if (value == null || value === "") return null;
  const numeric = Number(value);
  return Number.isFinite(numeric) ? numeric : null;
}

function makeReferenceId(label, seed) {
  return `reference_${label}_${stableHash(seed).toString(36)}_${Date.now().toString(36)}`;
}

function stableHash(value) {
  let hash = 2166136261;
  for (let index = 0; index < String(value).length; index += 1) {
    hash ^= String(value).charCodeAt(index);
    hash = Math.imul(hash, 16777619);
  }
  return hash >>> 0;
}

function basename(path) {
  return String(path ?? "").split(/[\\/]/).filter(Boolean).pop() ?? "";
}

File: shared/timeline/references.py
from __future__ import annotations

import math
import re
from copy import deepcopy
from typing import Any

from ..contracts.video_timeline import ASSET_SOURCE_FILE_PATH, ASSET_TYPE_IMAGE

REFERENCE_KIND_CHARACTER = "character"
REFERENCE_TAG_RE = re.compile(
    r"@(?P<label>image[1-9]\d*):(?P<kind>[A-Za-z][A-Za-z0-9_-]*)"
    r"(?:\[(?P<strength>[^\]]*)\])?"
)


def get_character_references(timeline: dict[str, Any]) -> list[dict[str, Any]]:
    references = (
        timeline.get("project", {})
        .get("metadata", {})
        .get("character_references", [])
    )
    return references if isinstance(references, list) else []


def are_character_references_enabled(timeline: dict[str, Any]) -> bool:
    return (
        timeline.get("project", {})
        .get("metadata", {})
        .get("character_references_enabled")
        is not False
    )


def normalize_character_references(value: Any) -> list[dict[str, Any]]:
    if not isinstance(value, list):
        return []

    normalized: list[dict[str, Any]] = []
    for index, reference in enumerate(value):
        if not isinstance(reference, dict):
            continue
        item = deepcopy(reference)
        item["label"] = normalize_reference_label(item.get("label"), len(normalized))
        item["id"] = str(item.get("id") or item["label"])
        item["kind"] = REFERENCE_KIND_CHARACTER
        item["enabled"] = item.get("enabled") is not False
        item["description"] = str(item.get("description") or "")
        item["strength"] = _clamp_strength(item.get("strength"))
        item["image"] = normalize_reference_image(item.get("image"))
        normalized.append(item)
    return normalized


def normalize_reference_label(value: Any, fallback_index: int = 0) -> str:
    label = str(value or "").strip().lower()
    if re.fullmatch(r"image[1-9]\d*", label):
        return label
    return f"image{fallback_index + 1}"


def normalize_reference_image(value: Any) -> dict[str, Any] | None:
    if not isinstance(value, dict):
        return None
    image = deepcopy(value)
    path = str(image.get("path") or image.get("file_path") or "").strip()
    image["type"] = ASSET_TYPE_IMAGE
    image["source_kind"] = image.get("source_kind") or ASSET_SOURCE_FILE_PATH
    image["path"] = path or None
    image["name"] = str(image.get("name") or _basename(path) or "")
    image["mime_type"] = image.get("mime_type") or ""
    image["size_bytes"] = image.get("size_bytes") if isinstance(image.get("size_bytes"), int) else None
    if not isinstance(image.get("metadata"), dict):
        image["metadata"] = {}
    image.pop("asset_id", None)
    return image


def parse_reference_tags(prompt: Any) -> list[dict[str, Any]]:
    tags: list[dict[str, Any]] = []
    for match in REFERENCE_TAG_RE.finditer(str(prompt or "")):
        kind = str(match.group("kind") or "").lower()
        tags.append(
            {
                "label": normalize_reference_label(match.group("label")),
                "kind": kind,
                "token": match.group(0),
                "supported": kind == REFERENCE_KIND_CHARACTER,
                "strength_override": _finite_float_or_none(match.group("strength")),
            }
        )
    return tags


def _clamp_strength(value: Any) -> float:
    try:
        numeric = float(value)
    except (TypeError, ValueError):
        return 1.0
    if not math.isfinite(numeric):
        return 1.0
    return max(0.0, min(1.0, numeric))


def _finite_float_or_none(value: Any) -> float | None:
    try:
        numeric = float(value)
    except (TypeError, ValueError):
        return None
    return numeric if math.isfinite(numeric) else None


def _basename(path: Any) -> str:
    return str(path or "").replace("\\", "/").rstrip("/").split("/")[-1]
